openshift/distro/ubi/9/Dockerfile

```diff
@@ -28,80 +28,53 @@
- && bundle config --local without ${BUNDLE_WITHOUT} \
- && bundle config --local gemfile ${BUNDLE_GEMFILE}
+RUN microdnf module enable -y "ruby:${RUBY_VERSION}" \
+ && microdnf install -y --nodocs ${BUILD_DEPS}
 
 # Copy sources
-COPY --chown=1001:1001 ./ "${HOME}/app/"
+WORKDIR /opt/source
+COPY ./ ./
 
-# Builds a clean source tree and deploys it with Bundler.
-# Sets the right configuration and permissions.
+# setup system
 RUN cp -n openshift/3scale_backend.conf /etc/ \
- && chmod 644 /etc/3scale_backend.conf \
- && BACKEND_VERSION=$(gem build apisonator.gemspec | \
-      sed -n -e 's/^\s*Version\:\s*\([^[:space:]]*\)$/\1/p') \
- && gem unpack "apisonator-${BACKEND_VERSION}.gem" --target=/opt/ruby \
- && cd "/opt/ruby/apisonator-${BACKEND_VERSION}" \
- && cp -a ${HOME}/app/.bundle "/opt/ruby/apisonator-${BACKEND_VERSION}/" \
- && bundle config --local path "/opt/ruby/apisonator-${BACKEND_VERSION}/" \
- && bundle install --jobs $(grep -c processor /proc/cpuinfo) \
- && ln -s ${PWD} /opt/app \
- && cp ${HOME}/app/openshift/config/puma.rb ./config/ \
- && cp -n ${HOME}/app/openshift/backend-cron /usr/local/sbin/backend-cron \
- && cp -n ${HOME}/app/openshift/entrypoint.sh ./ \
- && rm -rf ${HOME}/app \
+ && chmod 660 /etc/3scale_backend.conf \
+ && cp -n openshift/backend-cron /usr/local/sbin/backend-cron \
  && mkdir -p -m 0770 /var/run/3scale/ \
  && mkdir -p -m 0770 /var/log/backend/ \
  && touch /var/log/backend/3scale_backend{,_worker}.log \
- && chmod g+rw /var/log/backend/3scale_backend{,_worker}.log
-
-# Ensure safe defaults within the container for general case as well as to cover OpenShift specific case where the unprivileged user who runs the container is member of the root Linux group
-RUN chown -R 1001:0 /opt/ruby \
-    && chmod -R 750 /opt/ruby
-
+ && chmod g+rw /var/log/backend/3scale_backend{,_worker}.log \
+ && chown -R 1001 . && chmod g+w /opt
 
-EXPOSE 3000
+# Install bundler globally (as root)
+RUN BUNDLED_WITH=$(awk '/BUNDLED WITH/ { getline; print $1 }' Gemfile.lock) \
+ && gem install --no-document bundler:$BUNDLED_WITH \
+ && echo Using $(bundle --version), originally bundled with ${BUNDLED_WITH}
 
 USER 1001
 
+# Builds a clean source tree and deploys it with Bundler.
+RUN BACKEND_VERSION=$(gem build apisonator.gemspec | \
+      sed -n -e 's/^\s*Version\:\s*\([^[:space:]]*\)$/\1/p') \
+ && gem unpack "apisonator-${BACKEND_VERSION}.gem" --target=/tmp/ruby \
+ && mv "/tmp/ruby/apisonator-${BACKEND_VERSION}" /opt/app \
+ && rm -rf /tmp/ruby
+
 WORKDIR /opt/app
 
+# Install ruby dependencies and put things at the right places
+RUN bundle config set --local build.nokogiri --use-system-libraries \
+ && bundle config set --local without development:test \
+ && bundle config set --local deployment true \
+ && bundle install --jobs $(grep -c processor /proc/cpuinfo) \
+ && cp /opt/source/openshift/config/puma.rb config/ \
+ && cp -n /opt/source/openshift/entrypoint.sh ./ \
+ && rm -rf /opt/source
```

Comment:
This will make `/opt/source` unavailable in the final image, but the folder and its contents will continue to exist in previous layer and contribute to the final image size. The solution would be using a multi-stage dockerfile, but I think it's not worth it since this is a development dockerfile and probably `/opt/source` won't take too much space anyway.